### PR TITLE
fuse batch norm for conv operator 

### PR DESCRIPTION
**Pull request by luotao1**
related #9629 
- solve the fuse batch norm for conv operator with and without bias
- after fusing, the elapsed time on resnet (test_inference_image_classification) is **from 11.2s to 9.3s, about 10% speedup** on inference.

Note that this PR modify the program desc from C++ end, and discussed with @jacquesqiao, it is better to modify the program from Python end.

### CODE REVIEW

File: python/paddle/fluid/inference_transpiler.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        bias_op = self.block.insert_op(
+            index,
+            type="elementwise_add",
+            inputs={"X": x_var,
+                    "Y": y_var},
+            outputs={"Out": out_var},
+            attrs={"axis": 1})  # dim_start=1
+        return bias_op
+
+    def _fuse_param(self, current_op, bn_op, bias_op, with_bias):
```
[the remark this answers is not in the document]
Comment (NHZlX) [reply python/paddle/fluid/inference_transpiler.py]: @Xreki  原来的program是指什么呢？ 这里merge bn后，可以不用考虑原来的program是什么样子的吧，只要保证现在merge 后的正确就可以的。

File: python/paddle/fluid/inference_transpiler.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        self.place = place
+        self.block = program.block(0)
+        self.input_map = {}  # store the input names should be adjusted 
+
+        i = 0
+        while i < len(self.block.ops):
+            current_op = self.block.ops[i]
+            # TODO(luotao1): consider only conv2d now. fc would be delt later.
+            if current_op.type in ['conv2d']:
+                next_op = self.block.ops[i + 1]
```
[the remark this answers is not in the document]
Comment (NHZlX) [reply python/paddle/fluid/inference_transpiler.py]: @Xreki  可以详细描述一下么， 为什么获取到的next_op 是不正确的呢

**Comment by NHZlX** [comment]
LGTM

File: python/paddle/fluid/inference_transpiler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Xreki) [reply python/paddle/fluid/inference_transpiler.py]: 对于这种有分支的网络，some op的next op应该有两个，并且这种next op的取法依赖于op的保存顺序。batch_norm op几乎不会出现在这种分叉处，暂时可以先不用处理，加个comment说明一下吧。

File: python/paddle/fluid/inference_transpiler.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Xreki) [reply python/paddle/fluid/inference_transpiler.py]: @NHZlX 用户的行为难以预料，一旦用户clone了一个program的备份，并且希望后面继续使用这个program，就可能会出错。

File: python/paddle/fluid/inference_transpiler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (luotao1) [reply python/paddle/fluid/inference_transpiler.py]: Done

File: python/paddle/fluid/inference_transpiler.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+# limitations under the License.
+
+import numpy as np
+import os
+import shutil
+from . import core
+
+
+class InferenceTranspiler:
+    def transpile(self, program, scope, place):
```
Comment (Xreki) [inline python/paddle/fluid/inference_transpiler.py]: - 调整一下参数的顺序吧，变成：
```python
def transpile(self, program, place, scope=None):
```

- 当用户传入参数`scope`为`None`时，可使用默认[scope](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/executor.py#L27)。
- 检查一下三个参数的类型

File: python/paddle/fluid/inference_transpiler.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+                        self.block.remove_op(i + 2)
+                        i = i + 1
+            i = i + 1
+
+        self._adjust_input()
+        self._remove_unused_var()
+        # TODO(luotao): use clone() method to flush the program.desc in force, 
+        # since some large program.desc will not be flushed immediately. 
+        # And a better solution will be considered later.
+        program = program.clone()
```
Comment (Xreki) [inline python/paddle/fluid/inference_transpiler.py]: 这个`clone`我不确定是不是还是必须的，我特意把这里注释了，单独跑例子也没有出现问题。不过这里多一遍`clone`也没有关系。

File: python/paddle/fluid/inference_transpiler.py (hunk 3 above)
Comment (luotao1) [reply python/paddle/fluid/inference_transpiler.py]: Done
